engine/resource/src/resource.cpp

```diff
@@ -1038,28 +1042,108 @@ static const char* GetExtFromPath(const char* name, char* buffer, uint32_t buffe
+
+        ResourceCreateParams params;
+        params.m_Factory     = factory;
+        params.m_Context     = resource_type->m_Context;
+        params.m_Buffer      = buffer;
+        params.m_BufferSize  = buffer_size;
+        params.m_PreloadData = preload_data;
+        params.m_Resource    = &tmp_resource;
+        params.m_Filename    = name;
+        create_error         = resource_type->m_CreateFunction(params);
+    }
+
+    if (create_error == RESULT_OK && resource_type->m_PostCreateFunction)
+    {
+        ResourcePostCreateParams params;
+        params.m_Factory     = factory;
+        params.m_Context     = resource_type->m_Context;
+        params.m_PreloadData = preload_data;
+        params.m_Resource    = &tmp_resource;
+        for(;;)
+        {
+            create_error = resource_type->m_PostCreateFunction(params);
+            if(create_error != RESULT_PENDING)
+                break;
+            dmTime::Sleep(1000);
+        }
+    }
 
-    *resource = 0;
+    // Restore to default buffer size
+    factory->m_Buffer.SetSize(0);
+    if (factory->m_Buffer.Capacity() != DEFAULT_BUFFER_SIZE) {
+        factory->m_Buffer.SetCapacity(DEFAULT_BUFFER_SIZE);
+    }
 
-    char canonical_path[RESOURCE_PATH_MAX]; // The actual resource path. E.g. "/my/icon.texturec"
-    GetCanonicalPath(name, canonical_path);
+    if (create_error == RESULT_OK)
+    {
+        Result insert_error = InsertResource(factory, name, canonical_path_hash, &tmp_resource);
+        if (insert_error == RESULT_OK)
+        {
+            *resource_out = tmp_resource.m_Resource;
+            return RESULT_OK;
+        }
+        else
+        {
+            ResourceDestroyParams params;
+            params.m_Factory  = factory;
+            params.m_Context  = resource_type->m_Context;
+            params.m_Resource = &tmp_resource;
+            resource_type->m_DestroyFunction(params);
+            return insert_error;
+        }
+    }
+    else
+    {
+        dmLogWarning("Unable to create resource: %s: %s", canonical_path, ResultToString(create_error));
+        return create_error;
+    }
+}
 
-    uint64_t canonical_path_hash = dmHashBuffer64(canonical_path, strlen(canonical_path));
+// Assumes m_LoadMutex is already held
+static Result PrepareResourceCreation(HFactory factory, const char* canonical_path, dmhash_t canonical_path_hash, void** resource_out, SResourceType** resource_type_out)
+{
+    *resource_out = 0;
 
     // Try to get from already loaded resources
     SResourceDescriptor* rd = factory->m_Resources->Get(canonical_path_hash);
     if (rd)
     {
         assert(factory->m_ResourceToHash->Get((uintptr_t) rd->m_Resource));
         rd->m_ReferenceCount++;
-        *resource = rd->m_Resource;
-        return RESULT_OK;
+        *resource_out = rd->m_Resource;
+        return RESULT_ALREADY_REGISTERED;
```

Comment:
Note to self, maybe not throw error here. It's used for early outing in the calling functions, but we can maybe look at the resource pointer instead